### ISSUE TITLE
Using TypeVar, Generic and class methods can lead to Pyright incorrectly think there is a circular dependency 

### ISSUE THREAD

**Issue by Serhiy1** [microsoft/pyright#5086]
**Describe the bug**
I'm trying to add type hinting to a large and old project and I have encountered a situation where Pyright incorrectly reports an error of `Class definition for "Container" depends on itself`. Running the code shows no circular dependencies, as forward references were used. 

Environment:
Linux, python 3.8, pylance v2023.5.10    

**To Reproduce**
Here is a snippet of code that reproduces the error 

``` python
from typing import TYPE_CHECKING, Any, Generic, Type, TypeVar


item_one_type = TypeVar("item_one_type", bound="ItemOne")
item_two_type = TypeVar("item_two_type", bound="ItemTwo")
T = TypeVar("T")


class BaseContainer(Generic[item_one_type, item_two_type]):

    SetAtRuntime = Any
    if TYPE_CHECKING:
        ItemOneClass: Type[item_one_type] = SetAtRuntime
        ItemTwoClass: Type[item_two_type] = SetAtRuntime
    else:
        ItemOneClass = None
        ItemTwoClass = None

    @classmethod
    def register(cls, item_class: Type[T]) -> Type[T]:
        """ Subscribe Method for the class, basically sets the class variable ItemX when
        used as a decorator
        """
        name = f"{item_class.__name__}Class"
        if not hasattr(cls, name):
            raise TypeError(f"Cannot register '{name}'.")
        setattr(cls, name, item_class)
        return item_class

    def create_item_one(self, *args, **kwargs) -> ItemOneClass:
        if self.ItemOneClass is None:
            raise Exception("ItemOne is Unknown")
        return self.ItemOneClass(*args, **kwargs)

    def create_item_two(self, *args, **kwargs) -> ItemTwoClass:
        if self.ItemTwoClass is None:
            raise Exception("ItemOne is Unknown")
        return self.ItemTwoClass(*args, **kwargs)


class Container(BaseContainer["ItemOne", "ItemTwo"]):
    ...


# There are multiple definitions of ItemOne and ItemTwo in different parts of 
# the codebase, hence the use of the observer pattern

@Container.register     # Error is recorded here
class ItemTwo:

    def __init__(self, x: int, y: str, z: int) -> None:
        self.x = x
        self.y = y
        self.z = z


@Container.register     # Error is recorded here
class ItemOne:
    def __init__(self, x: int, y: str) -> None:
        self.x = x
        self.y = y


Container().create_item_one(2, "str")      # works during runtime
Container().create_item_two(2, "str", 3)    # works during runtime

```

**Expected Behaviour**
I would expect Pyright to recognise that forward references were used in the `BaseContainer` and `Container` definitions and not throw an error

**Comment by erictraut** [comment]
This is by design. The `Container` class uses a decorated class called `ItemTwo` in its class hierarchy, but `ItemTwo` is decorated with an expression that uses `Container`. This is a true circular dependency. You'll need to eliminate the circularity if you want to avoid the error in pyright.
